WWDC/DownloadManager.swift

```diff
@@ -178,109 +177,102 @@ final class DownloadManager: NSObject, Logging {
+                    return .downloading(latestInfo)
+                case .suspended:
+                    return .paused(latestInfo)
+                case .canceling:
+                    return .cancelled
+                case .completed:
+                    return .finished
+                @unknown default:
+                    assertionFailure("An unexpected case was discovered on an non-frozen obj-c enum")
+                    return .downloading(latestInfo)
                 }
+            } else if self.hasDownloadedVideo(remoteURL: asset.remoteURL) {
+                return .finished
+            } else {
+                return .none
             }
+        }
 
-            checkDownloadedState()
-
-            let fileDeleted = nc.dm_addObserver(forName: .DownloadManagerFileDeletedNotification, filteredBy: asset.relativeLocalURL) { _ in
-
-                observer.onNext(.none)
-            }
-
-            let fileAdded = nc.dm_addObserver(forName: .DownloadManagerFileAddedNotification, filteredBy: asset.relativeLocalURL) { _ in
-
-                observer.onNext(.finished)
-            }
-
-            let started = nc.dm_addObserver(forName: .DownloadManagerDownloadStarted, filteredBy: asset.remoteURL) { _ in
-
-                observer.onNext(.downloading(.unknown))
-            }
-
-            let cancelled = nc.dm_addObserver(forName: .DownloadManagerDownloadCancelled, filteredBy: asset.remoteURL) { _ in
-
-                observer.onNext(.cancelled)
-            }
-
-            let paused = nc.dm_addObserver(forName: .DownloadManagerDownloadPaused, filteredBy: asset.remoteURL) { _ in
-
-                observer.onNext(.paused(latestInfo))
-            }
-
-            let resumed = nc.dm_addObserver(forName: .DownloadManagerDownloadResumed, filteredBy: asset.remoteURL) { _ in
-
-                observer.onNext(.downloading(latestInfo))
-            }
-
-            let failed = nc.dm_addObserver(forName: .DownloadManagerDownloadFailed, filteredBy: asset.remoteURL) { note in
+        let nc = NotificationCenter.default
+        let fileDeleted = nc.publisher(for: .DownloadManagerFileDeletedNotification, filteredBy: asset.relativeLocalURL).map { _ in
+            DownloadStatus.none
+        }
+        let fileAdded = nc.publisher(for: .DownloadManagerFileAddedNotification, filteredBy: asset.relativeLocalURL).map { _ in
+            DownloadStatus.finished
+        }
 
-                let error = note.userInfo?["error"] as? Error
-                observer.onNext(.failed(error))
+        let progress = nc.publisher(for: .DownloadManagerDownloadProgressChanged, filteredBy: asset.remoteURL).map { note in
+            if let info = note.userInfo?["info"] as? DownloadInfo {
+                latestInfo = info
+                if info.taskState == .suspended {
+                    // We can get progress updates that were from while the task was suspending
+                    return DownloadStatus.paused(info)
```

Comment:
@insidegui This fixes the bug where the pause/resume button gets into the wrong state which has been there since I implemented it like 4 years ago!